driver/src/main/java/org/neo4j/driver/internal/async/pool/ConnectionPoolImpl.java

```diff
@@ -156,37 +165,40 @@ public int idleConnections( BoltServerAddress address )
     {
         if ( closed.compareAndSet( false, true ) )
         {
-            try
-            {
-                nettyChannelTracker.prepareToCloseChannels();
-                for ( Map.Entry<BoltServerAddress,ExtendedChannelPool> entry : pools.entrySet() )
-                {
-                    BoltServerAddress address = entry.getKey();
-                    ExtendedChannelPool pool = entry.getValue();
-                    log.info( "Closing connection pool towards %s", address );
-                    closePool( pool );
-                }
+            nettyChannelTracker.prepareToCloseChannels();
 
-                pools.clear();
-            }
-            finally
-            {
+            CompletableFuture<Void> allPoolClosedFuture = CompletableFuture.allOf(
+                    pools.entrySet().stream().map( entry -> {
+                        BoltServerAddress address = entry.getKey();
+                        ExtendedChannelPool pool = entry.getValue();
+                        log.info( "Closing connection pool towards %s", address );
+                        // Wait for all pools to be closed.
+                        return closePool( pool ).toCompletableFuture();
+                    } ).toArray( CompletableFuture[]::new ) );
 
-                if (ownsEventLoopGroup) {
+            // We can only shutdown event loop group when all netty pools are fully closed,
+            // otherwise the netty pools might missing threads (from event loop group) to execute clean ups.
+            allPoolClosedFuture.whenComplete( ( ignored, pollCloseError ) -> {
+                pools.clear();
+                if ( !ownsEventLoopGroup )
+                {
+                    completedWithNullIfNonError( closeFuture, pollCloseError );
+                }
+                else
+                {
                     // This is an attempt to speed up the shut down procedure of the driver
                     // Feel free return this back to shutdownGracefully() method with default values
                     // if this proves troublesome!!!
-                    eventLoopGroup().shutdownGracefully(200, 15_000, TimeUnit.MILLISECONDS);
+                    eventLoopGroup().shutdownGracefully( 200, 15_000, TimeUnit.MILLISECONDS );
```

Comment:
Maybe extract all event loop shutdown logic in a separate method? (both `shutdownGracefully` and `whenComplete` parts)